compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -960,47 +989,327 @@ class JSCodeGen()(using genCtx: Context) {
 
   // Constructor of a non-native JS class ------------------------------------
 
-  def genJSClassCapturesAndConstructor(classSym: Symbol,
-      constructorTrees: List[DefDef]): (Option[List[js.ParamDef]], js.JSMethodDef) = {
-    implicit val pos = classSym.span
+  def genJSClassCapturesAndConstructor(constructorTrees: List[DefDef])(
+      implicit pos: SourcePosition): (List[js.ParamDef], js.JSMethodDef) = {
+    /* We need to merge all Scala constructors into a single one because the
+     * IR, like JavaScript, only allows a single one.
+     *
+     * We do this by applying:
+     * 1. Applying runtime type based dispatch, just like exports.
+     * 2. Splitting secondary ctors into parts before and after the `this` call.
+     * 3. Topo-sorting all constructor statements and including/excluding
+     *    them based on the overload that was chosen.
+     */
 
-    if (hasDefaultCtorArgsAndJSModule(classSym)) {
-      report.error(
-          "Implementation restriction: " +
-          "constructors of non-native JS classes cannot have default parameters if their companion module is JS native.",
-          classSym.srcPos)
-      val ctorDef = js.JSMethodDef(js.MemberFlags.empty,
-          js.StringLiteral("constructor"), Nil, None, js.Skip())(
-          OptimizerHints.empty, None)
-      (None, ctorDef)
-    } else {
-      withNewLocalNameScope {
-        localNames.reserveLocalName(JSSuperClassParamName)
+    val (primaryTree :: Nil, secondaryTrees) =
+      constructorTrees.partition(_.symbol.isPrimaryConstructor)
 
-        val ctors: List[js.MethodDef] = constructorTrees.flatMap { tree =>
-          genMethodWithCurrentLocalNameScope(tree)
-        }
+    val primaryCtor = genPrimaryJSClassCtor(primaryTree)
+    val secondaryCtors = secondaryTrees.map(genSecondaryJSClassCtor(_))
 
-        val (captureParams, dispatch) =
-          jsExportsGen.genJSConstructorDispatch(constructorTrees.map(_.symbol))
+    // VarDefs for the parameters of all constructors.
+    val paramVarDefs = for {
+      vparam <- constructorTrees.flatMap(_.paramss.flatten)
+    } yield {
+      val sym = vparam.symbol
+      val tpe = toIRType(sym.info)
+      js.VarDef(encodeLocalSym(sym), originalNameOfLocal(sym), tpe, mutable = true, jstpe.zeroOf(tpe))(vparam.span)
```

Comment:
I noticed that the corresponding scala-js code uses more linebreaks in expressions (for example this line is split in two in scala-js), wouldn't it be better to mimic the scala-js style to simplify comparisons?

Reply:
Yes, perhaps that would be better. Is it worth spending my time reformatting things as close as possible to the scala-js code?

Reply:
> Is it worth spending my time reformatting things

No, since you're maintaining this code you get to do decide how to format it :).